spring-kafka/src/test/java/org/springframework/kafka/listener/KafkaMessageListenerContainerTests.java

```diff
@@ -661,8 +661,25 @@ public void testBatchListenerManual() throws Exception {
 				containerProps);
 		container.setBeanName("testBatchListenerManual");
 		container.start();
+		Consumer<?, ?> containerConsumer = spyOnConsumer(container);
+		final CountDownLatch commitLatch = new CountDownLatch(2);
+		willAnswer(invocation -> {
+
+			@SuppressWarnings({ "unchecked" })
+			Map<TopicPartition, OffsetAndMetadata> map = (Map<TopicPartition, OffsetAndMetadata>) invocation
+					.getArguments()[0];
+			for (Entry<TopicPartition, OffsetAndMetadata> entry : map.entrySet()) {
+				if (entry.getValue().offset() == 2) {
+					commitLatch.countDown();
+				}
+			}
+			return invocation.callRealMethod();
```

Comment:
Well, I don't say that I was so smart that time. And I am still not sure that I am such right now or will be tomorrow.
There is really some magic around that makes some explosions in a head and after that it needs abundant firefighting. :smile: